### PR TITLE
allow object to remove previous class, also dedupes class names

### LINKED ISSUE #18: Allow replacing of previous class names
My use case is something like `classNames('foo', { foo: false, bar: true })` should render out `class="bar"` instead of `class="foo bar"`. Should this be supported by the library or is this out of scope? I'd be glad to submit a PR if this use case is something you guys plan to support.

### PR DESCRIPTION
**Pull request by longlho**
fix #18 @JedWatson

### CODE REVIEW

**Comment by JedWatson** [comment]
@longlho thanks for this!

Based on what we learned with #16 I am concerned it will reduce performance dramatically. Would you mind setting up a JSPerf to test? You can use http://jsperf.com/classnames-util/3 as a base.

I'll add a contributing.md with a perf concerns note as well so this priority is clearer in the future.

**Comment by dcousens** [comment]
@longlho maybe split this into 2 PRs?

The tests are great! :)

**Comment by longlho** [comment]
I think that having the output of `foo foo bar` as `class` is not the best, but it is not wrong. I think in terms of correctness it'd be best to de-dupe.

@JedWatson it does take a performance hit if we're talking about using this 1000s of times over. I'll make a `jsperf` but with the typical use case of having even 10s of classes, it shouldn't make much of a difference IMHO.

I personally can't think of a better solution than map conversion since lookup is constant :( Honestly, again if we're talking about <100 classes, linear lookup might even cost less than setting up data structures.

**Comment by JedWatson** [comment]
@longlho the issue with perf is that if you're using this in across hundreds of React components across an app (easily feasible), each render call will hit it to generate the new class string resulting in hundreds if not thousands of executions per render.

So even when it's only processing a few classes for each call, it should be as fast as possible (including not just the performance of data manipulations but memory allocation, etc.)

The map solution makes sense but requires splitting strings into arrays etc. - it might be worth running a few comparisons to test against stripping w/ a RegExp for false values. The simplicity of surrounding code may make up for the non-linear performance of the string lookup. See http://jsperf.com/string-replace-str-vs-regexp

You could combine the two concepts by adding classes to an internal map _as well as the string_ so you can check with a lookup to see if they need to be removed.

As per @dcousens's comment on #18 though I'm not sure this is really in scope for the package, seems like a lot of additional complexity that could be resolved higher-up in the implementation...

**Comment by longlho** [comment]
@JedWatson hmm gotcha I'll try w/ string replace since it's likely to be a small set of classes. The de-dupe is just a nice-to-have w/ the map conversion solution so if we go w/ string replace and skip de-duping I'm cool w/ that :)

**Comment by JedWatson** [comment]
@longlho two ideas:

1. maybe we make this available via a separate entry point, so that the more complex logic is "opt in" if you need it

```js
var classNames = require('classNames');
// ...
<div className={classNames.dedupe('foo', { foo: false, bar: true })} />
```

2. maybe we can keep an internal map of classes added, and then only do the replace when we know we have to. peseudo logic for the example above:

```js
// first argument
cache['foo'] = true;
rtn = 'foo ';
// second argument, first key in obj
if (cache['foo']) {
  rtn = rtn.replace('foo ', '');
  delete cache['foo'];
}
```

sorry for the dodgy example, hope that makes my suggestion clear enough

**Comment by longlho** [comment]
@JedWatson gotcha! I'll give that a spin

**Comment by longlho** [comment]
@JedWatson I gave that a spin and the string replace is indeed faster. Data structure creation might be creating too much overhead for such a small number of classes. I added `benchmarks.js` & travis integration as well.

File: index.js [hunk 1]
```diff
@@ -1,29 +1,67 @@
...
-			for (var key in arg) {
-				if (!arg.hasOwnProperty(key) || !arg[key]) {
+			for (k in arg) {
+				if (!arg.hasOwnProperty(k)) {
 					continue;
 				}
-				classes += ' ' + key;
+				v = arg[k];
+				if (!v) {
+					classes = classes.replace(new RegExp(' ' + k, 'g'), '');
```
Comment (cezar-berea) [inline index.js]: You need a better regexp. This is going to fail this test:
```js
assert.equal(classnames("foo", "foobar", "bar", {foo: false}), "foobar")); 
# yields "ar bar" instead of "foobar bar"
```

**Comment by longlho** [comment]
@cezar-berea @JedWatson fixed the regex bug

**Comment by longlho** [comment]
ping @JedWatson @dcousens

**Comment by JedWatson** [comment]
Sorry for the delay @longlho had a crazy week at work and want to give this some real attention. I'll pick it up this weekend :)

**Comment by longlho** [comment]
ping @dcousens & @JedWatson again :)

**Comment by longlho** [comment]
ping @dcousens & @JedWatson. If the change is not accepted u guys can just close this PR.

**Comment by dcousens** [comment]
Waiting on a conclusion in #18.  Sorry to keep you waiting @longlho

**Comment by dcousens** [comment]
ping @JedWatson

File: tests.js [hunk 2]
```diff
@@ -1,60 +1,93 @@
...
+			f: 1,
+		}), 'a f');
+	});
+
+	it('should dedupe classNames', function () {
+		assert(_hasSameClasses(classNames('foo', 'bar', 'foo', 'bar', { foo: true }), 'foo bar'));
+	});
+
+	it('should make sure subsequent objects can remove/add classes', function () {
+		assert(_hasSameClasses(classNames('foo', { foo: false }, { foo: true, bar: true }), 'foo bar'));
```
Comment (3den) [inline tests.js]: :+1:

File: tests.js [hunk 3]
```diff
@@ -1,60 +1,93 @@
...
+			b: false,
+			c: 0,
+			d: null,
+			e: undefined,
+			f: 1,
+		}), 'a f');
+	});
+
+	it('should dedupe classNames', function () {
+		assert(_hasSameClasses(classNames('foo', 'bar', 'foo', 'bar', { foo: true }), 'foo bar'));
```
Comment (3den) [inline tests.js]: :+1:

File: tests.js [hunk 4]
```diff
@@ -1,60 +1,93 @@
...
+	});
+
+	it('should make sure subsequent objects can remove/add classes', function () {
+		assert(_hasSameClasses(classNames('foo', { foo: false }, { foo: true, bar: true }), 'foo bar'));
+	});
+
+	it('should make sure object with falsy value wipe out previous classes', function () {
+		assert(_hasSameClasses(classNames('foo foo', 0, null, undefined, true, 1, 'b', { 'foo': false }), '1 b'));
```
Comment (3den) [inline tests.js]: we should not need this `_hasSameClasses` helper since it is not providing a custom assertion nor abstracting the call. Also I am not sure if the test pass because _hasSameClasses or because classNames actually works as expected. A simple `assert.equal` would be much better.

File: tests.js [hunk 5]
```diff
@@ -1,60 +1,93 @@
...
+	it('handles arrays that include arrays', function() {
+		assert(_hasSameClasses(classNames(['a', ['b', 'c']]), 'a b c'));
+	});
+
+	it('handles arrays that include objects', function() {
+		assert(_hasSameClasses(classNames(['a', {b: true, c: false}]), 'a b'));
+	});
+
+	it('handles deep array recursion', function() {
+		assert(_hasSameClasses(classNames(['a', ['b', ['c', {d: true}]]]), 'a b c d'));
```
Comment (3den) [inline tests.js]: wow, i am not sure if someone would ever need this

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+language: node_js
+node_js:
+  - "0.12"
```
Comment (dcousens) [inline .travis.yml]: please put in a separate commit so we can cherry-pick, or put it in another PR

File: .jshintrc [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+	"camelcase": true
```
Comment (dcousens) [inline .jshintrc]: This is not relevant to the PR

File: benchmarks.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
```
Comment (dcousens) [inline benchmarks.js]: Nice bench code, but lets use the merged benchmark script now.

File: index.js [hunk 9]
```diff
@@ -1,30 +1,60 @@
...
 
-// safely export classNames in case the script is included directly on a page
-if (typeof module !== 'undefined' && module.exports) {
-	module.exports = classNames;
-}
+	// safely export classNames in case the script is included directly on a page
+	if (typeof module !== 'undefined' && module.exports) {
+		module.exports = classNames;
+	} else if (typeof window !== 'undefined') {
+		window.classNames = classNames;
```
Comment (dcousens) [inline index.js]: @longlho please remove this.

File: tests.js (hunk 5 above)
Comment (dcousens) [inline tests.js]: Its rudimentary to keep,  so I'm happy @longlho tested it.

**Comment by dcousens** [comment]
@JedWatson I'm happy with this PR once the nits have been fixed.

* Use the now merged benchmark code,
* Remove the `windows.classname` export,
* Remove the `.jshintrc`,

I think that will get us to the point where we can merge this and make amendments where necessary.

**Comment by dcousens** [comment]
Also, @longlho please rebase on master.   I'll rebase it manually, fix nits and re-submit it in a day or 2 if otherwise necessary.

File: tests.js (hunk 4 above)
Comment (longlho) [inline tests.js]: @3den  the reason for `_hasSameClasses` is bc `object` key value are unordered, thus `assert.equal` won't suffice

**Comment by longlho** [comment]
@dcousens @JedWatson changed :)

**Comment by dcousens** [comment]
@JedWatson LGTM. +1.  We'll break it out as discussed after merge.

File: tests.js [hunk 10]
```diff
@@ -58,3 +60,94 @@ describe('classNames', function() {
...
+	it('handles arrays that include falsy and true values', function() {
+		assert(_hasSameClasses(dedupe(['a', 0, null, undefined, false, true, 'b']), 'a b'));
+	});
+
+	it('handles arrays that include arrays', function() {
+		assert(_hasSameClasses(dedupe(['a', ['b', 'c']]), 'a b c'));
+	});
+
+	it('handles arrays that include objects', function() {
+		assert(_hasSameClasses(dedupe(['a', {b: true, c: false}]), 'a b'));
```
Comment (3den) [inline tests.js]: there is to much code duplication ( `assert(_hasSameClasses(dedupe`) it would be be nice to have a custom assertion in this case, something like `assertDedupe(['a', {b: true, c: false}], 'a b')`...

Comment (3den) [inline tests.js]: the advantage of adding a custom assertion is because it is easy to add a meaningful error message to every test that uses that custom assertion and makes to code easier to change

**Comment by JedWatson** [comment]
Thanks @longlho, looks great. Merging now.

File: tests.js (hunk 10 above)
Comment (JedWatson) [inline tests.js]: :+1: sounds good @3den, I really like this idea. Are you interested in creating a PR?

**Comment by longlho** [comment]
comply
